samples/matter/common/src/bridge/bridge_manager.cpp

```diff
@@ -31,96 +31,149 @@ void BridgeManager::Init()
 				     false);
 }
 
-CHIP_ERROR BridgeManager::AddBridgedDevices(BridgedDevice *aDevice, BridgedDeviceDataProvider *aDataProvider)
+CHIP_ERROR BridgeManager::AddBridgedDevices(MatterBridgedDevice *devices[], BridgedDeviceDataProvider *dataProvider,
+					    uint8_t deviceListSize)
 {
-	aDataProvider->Init();
-	CHIP_ERROR err = AddDevices(aDevice, aDataProvider);
-
-	return err;
+	VerifyOrReturnError(deviceListSize <= kMaxDevicesPerProvider, CHIP_ERROR_NO_MEMORY);
+	dataProvider->Init();
+	return AddDevices(devices, dataProvider, deviceListSize);
 }
 
 CHIP_ERROR BridgeManager::RemoveBridgedDevice(uint16_t endpoint)
 {
 	uint8_t index = 0;
-
 	while (index < kMaxBridgedDevices) {
 		if (mDevicesMap.Contains(index)) {
-			const DevicePair *devicesPtr = mDevicesMap[index];
-			if (devicesPtr->mDevice->GetEndpointId() == endpoint) {
+			auto &devicePair = mDevicesMap[index];
+			if (devicePair.mDevice->GetEndpointId() == endpoint) {
 				LOG_INF("Removed dynamic endpoint %d (index=%d)", endpoint, index);
 				/* Free dynamically allocated memory */
 				emberAfClearDynamicEndpoint(index);
-				if (mDevicesMap.Erase(index)) {
-					mNumberOfProviders--;
-					return CHIP_NO_ERROR;
-				} else {
-					LOG_ERR("Cannot remove bridged devices under index=%d", index);
-					return CHIP_ERROR_NOT_FOUND;
-				}
+				return SafelyRemoveDevice(index);
 			}
 		}
 		index++;
 	}
 	return CHIP_ERROR_NOT_FOUND;
 }
 
-CHIP_ERROR BridgeManager::AddDevices(BridgedDevice *aDevice, BridgedDeviceDataProvider *aDataProvider)
+CHIP_ERROR BridgeManager::SafelyRemoveDevice(uint8_t index)
```

Comment:
Looking at the body of this function I can see it is used to remove duplications only, so maybe it is worth changing its name to RemoveDuplications or something like that.

Reply:
Actually this method checks if there is another Matter device (e.g. humidity) associated with the provider which we want to remove together with the Matter device under `index` (e.g. temperature). If so, then we cannot release the memory of this provider as it will be still used, so the implementation somehow mimics the shared pointer - we release the resource only when it stops being owned by all entities. Therefore the `RemoveDuplications` doesn't sound right here.